File: app/src/main/java/com/tungnui/dalatlaptop/ux/fragments/OrdersHistoryFragment.java
package com.tungnui.dalatlaptop.ux.fragments;

import android.app.Activity;
import android.app.ProgressDialog;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.tungnui.dalatlaptop.CONST;
import com.tungnui.dalatlaptop.MyApplication;
import com.tungnui.dalatlaptop.R;
import com.tungnui.dalatlaptop.entities.Metadata;
import com.tungnui.dalatlaptop.entities.order.Order;
import com.tungnui.dalatlaptop.interfaces.OrdersRecyclerInterface;
import com.tungnui.dalatlaptop.utils.EndlessRecyclerScrollListener;
import com.tungnui.dalatlaptop.utils.RecyclerMarginDecorator;
import com.tungnui.dalatlaptop.utils.Utils;
import com.tungnui.dalatlaptop.ux.MainActivity;
import com.tungnui.dalatlaptop.ux.adapters.OrdersHistoryRecyclerAdapter;
import timber.log.Timber;

/**
 * Fragment shows the user's order history.
 */
public class OrdersHistoryFragment extends Fragment {

    private ProgressDialog progressDialog;

    // Fields referencing complex screen layouts.
    private View empty;
    private View content;

    /**
     * Request metadata containing urls for endlessScroll.
     */
    private Metadata ordersMetadata;

    private OrdersHistoryRecyclerAdapter ordersHistoryRecyclerAdapter;
    private EndlessRecyclerScrollListener endlessRecyclerScrollListener;

    /**
     * Field for recovering scroll position.
     */
    private RecyclerView ordersRecycler;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        Timber.d("%s - onCreateView", this.getClass().getSimpleName());
        MainActivity.Companion.setActionBarTitle(getString(R.string.Order_history));

        View view = inflater.inflate(R.layout.fragment_orders_history, container, false);

        progressDialog = Utils.generateProgressDialog(getActivity(), false);

        empty = view.findViewById(R.id.order_history_empty);
        content = view.findViewById(R.id.order_history_content);

        prepareOrdersHistoryRecycler(view);

        loadOrders(null);
        return view;
    }

    /**
     * Prepare content recycler. Create custom adapter and endless scroll.
     *
     * @param view root fragment view.
     */
    private void prepareOrdersHistoryRecycler(View view) {
        ordersRecycler = (RecyclerView) view.findViewById(R.id.orders_history_recycler);
        ordersHistoryRecyclerAdapter = new OrdersHistoryRecyclerAdapter(new OrdersRecyclerInterface() {
            @Override
            public void onOrderSelected(View v, Order order) {
                Activity activity = getActivity();
                if (activity instanceof MainActivity) ((MainActivity) activity).onOrderSelected(order);
            }
        });
        ordersRecycler.setAdapter(ordersHistoryRecyclerAdapter);
        LinearLayoutManager layoutManager = new LinearLayoutManager(ordersRecycler.getContext());
        ordersRecycler.setLayoutManager(layoutManager);
        ordersRecycler.setItemAnimator(new DefaultItemAnimator());
        ordersRecycler.setHasFixedSize(true);
        ordersRecycler.addItemDecoration(new RecyclerMarginDecorator(getResources().getDimensionPixelSize(R.dimen.base_margin)));

        endlessRecyclerScrollListener = new EndlessRecyclerScrollListener(layoutManager) {
            @Override
            public void onLoadMore(int currentPage) {
                if (ordersMetadata != null && ordersMetadata.getLinks() != null && ordersMetadata.getLinks().getNext() != null) {
                    loadOrders(ordersMetadata.getLinks().getNext());
                } else {
                    Timber.d("CustomLoadMoreDataFromApi NO MORE DATA");
                }
            }
        };
        ordersRecycler.addOnScrollListener(endlessRecyclerScrollListener);
    }

    /**
     * Endless content loader. Should be used after views inflated.
     *
     * @param url null for fresh load. Otherwise use URLs from response metadata.
     */
    private void loadOrders(String url) {
      /*  User user = SettingsMy.getActiveUser();
        if (user != null) {
            progressDialog.show();
            if (url == null) {
                ordersHistoryRecyclerAdapter.clear();
                url = String.format(EndPoints.ORDERS, SettingsMy.getActualNonNullShop(getActivity()).getId());
            }
            GsonRequest<OrderResponse> req = new GsonRequest<>(Request.Method.GET, url, null, OrderResponse.class, new Response.Listener<OrderResponse>() {
                @Override
                public void onResponse(OrderResponse response) {
                    ordersMetadata = response.getMetadata();
                    ordersHistoryRecyclerAdapter.addOrders(response.getOrders());

                    if (ordersHistoryRecyclerAdapter.getItemCount() > 0) {
                        empty.setVisibility(View.GONE);
                        content.setVisibility(View.VISIBLE);
                    } else {
                        empty.setVisibility(View.VISIBLE);
                        content.setVisibility(View.GONE);
                    }
                    if (progressDialog != null) progressDialog.cancel();
                }
            }, new Response.ErrorListener() {
                @Override
                public void onErrorResponse(VolleyError error) {
                    if (progressDialog != null) progressDialog.cancel();
                    MsgUtils.logAndShowErrorMessage(getActivity(), error);
                }
            }, getFragmentManager(), user.getAccessToken());

            req.setRetryPolicy(MyApplication.getDefaultRetryPolice());
            req.setShouldCache(false);
            MyApplication.getInstance().addToRequestQueue(req, CONST.ORDERS_HISTORY_REQUESTS_TAG);
        } else {
            LoginExpiredDialogFragment loginExpiredDialogFragment = new LoginExpiredDialogFragment();
            loginExpiredDialogFragment.show(getFragmentManager(), "loginExpiredDialogFragment");
        }*/
    }

    @Override
    public void onStop() {
        if (progressDialog != null) {
            // Hide progress dialog if exist.
            if (progressDialog.isShowing() && endlessRecyclerScrollListener != null) {
                // Fragment stopped during loading data. Allow new loading on return.
                endlessRecyclerScrollListener.resetLoading();
            }
            progressDialog.cancel();
        }
        MyApplication.getInstance().cancelPendingRequests(CONST.ORDERS_HISTORY_REQUESTS_TAG);
        super.onStop();
    }

    @Override
    public void onDestroyView() {
        if (ordersRecycler != null) ordersRecycler.clearOnScrollListeners();
        super.onDestroyView();
    }
}

File: app/src/main/java/com/tungnui/dalatlaptop/ux/fragments/OrdersHistoryFragment.kt
package com.tungnui.dalatlaptop.ux.fragments

import android.app.ProgressDialog
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.tungnui.dalatlaptop.R
import com.tungnui.dalatlaptop.SettingsMy
import com.tungnui.dalatlaptop.api.EndPoints
import com.tungnui.dalatlaptop.api.ServiceGenerator
import com.tungnui.dalatlaptop.entities.Metadata
import com.tungnui.dalatlaptop.models.Order
import com.tungnui.dalatlaptop.utils.EndlessRecyclerScrollListener
import com.tungnui.dalatlaptop.utils.RecyclerMarginDecorator
import com.tungnui.dalatlaptop.utils.Utils
import com.tungnui.dalatlaptop.utils.getNextUrl
import com.tungnui.dalatlaptop.ux.MainActivity
import com.tungnui.dalatlaptop.ux.adapters.OrdersHistoryRecyclerAdapter
import com.tungnui.dalatlaptop.ux.dialogs.LoginExpiredDialogFragment
import com.tungnui.dalatlaptop.woocommerceapi.OrderServices
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.fragment_orders_history.*
import org.jetbrains.anko.support.v4.longToast
import org.jetbrains.anko.support.v4.toast
import timber.log.Timber

class OrdersHistoryFragment : Fragment() {
    private var mCompositeDisposable: CompositeDisposable
    val orderService: OrderServices
    init {
        mCompositeDisposable = CompositeDisposable()
        orderService = ServiceGenerator.createService(OrderServices::class.java)
    }
    private var progressDialog: ProgressDialog? = null

    /**
     * Request metadata containing urls for endlessScroll.
     */
    private val ordersMetadata: Metadata? = null
    private var order: Order? = null
    private var orderNextLink:String? = null

    private lateinit var ordersHistoryRecyclerAdapter: OrdersHistoryRecyclerAdapter
    private var endlessRecyclerScrollListener: EndlessRecyclerScrollListener? = null
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_orders_history, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        MainActivity.setActionBarTitle(getString(R.string.Order_history))
        progressDialog = Utils.generateProgressDialog(activity, false)
        prepareOrdersHistoryRecycler()
        loadOrders(null)
    }



    /**
     * Prepare content recycler. Create custom adapter and endless scroll.
     *
     * @param view root fragment view.
     */
    private fun prepareOrdersHistoryRecycler() {
        ordersHistoryRecyclerAdapter = OrdersHistoryRecyclerAdapter{
            val activity = activity
            (activity as? MainActivity)?.onOrderSelected(order)
        }
        orders_history_recycler.adapter = ordersHistoryRecyclerAdapter
        val layoutManager = LinearLayoutManager(orders_history_recycler.context)
        orders_history_recycler.layoutManager = layoutManager
        orders_history_recycler.itemAnimator = DefaultItemAnimator()
        orders_history_recycler.setHasFixedSize(true)
        orders_history_recycler.addItemDecoration(RecyclerMarginDecorator(resources.getDimensionPixelSize(R.dimen.base_margin)))

        endlessRecyclerScrollListener = object : EndlessRecyclerScrollListener(layoutManager) {
            override fun onLoadMore(currentPage: Int) {
                if (orderNextLink != null ) {
                    loadOrders(orderNextLink)
                } else {
                    Timber.d("CustomLoadMoreDataFromApi NO MORE DATA")
                }
            }
        }
        orders_history_recycler.addOnScrollListener(endlessRecyclerScrollListener)
    }

    private fun loadOrders(url: String?) {
        var user = SettingsMy.getActiveUser();
        var url:String? = url
        if (user != null) {
            progressDialog?.show();
            if (url == null) {
                ordersHistoryRecyclerAdapter.clear();
                url = EndPoints.ORDERS +"?customer=${user.id}"
            }
            var disposable = orderService.getAll(url)
                        .subscribeOn((Schedulers.io()))
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({ response ->
                            response.body()?.let { ordersHistoryRecyclerAdapter.addOrders(it) }
                            orderNextLink = response.headers()["Link"]?.getNextUrl()
                            if (ordersHistoryRecyclerAdapter.getItemCount() > 0) {
                                order_history_empty.setVisibility(View.GONE);
                                order_history_content.setVisibility(View.VISIBLE);
                            } else {
                                order_history_empty.setVisibility(View.VISIBLE);
                                order_history_content.setVisibility(View.GONE);
                            }
                            progressDialog?.cancel()
                        },
                                { error ->
                                    progressDialog?.cancel()
                                })

             mCompositeDisposable.add(disposable)
                   } else {
            var loginExpiredDialogFragment = LoginExpiredDialogFragment();
            loginExpiredDialogFragment.show(getFragmentManager(), "loginExpiredDialogFragment");
        }
    }

    override fun onStop() {
        if (progressDialog != null) {
            // Hide progress dialog if exist.
            if (progressDialog!!.isShowing && endlessRecyclerScrollListener != null) {
                // Fragment stopped during loading data. Allow new loading on return.
                endlessRecyclerScrollListener!!.resetLoading()
            }
            progressDialog!!.cancel()
        }
        super.onStop()
    }

    override fun onDestroyView() {
        orders_history_recycler.clearOnScrollListeners()
        super.onDestroyView()
    }
}
